src/dataflow-types/src/plan/mod.rs

```diff
@@ -397,6 +405,53 @@ impl Plan {
                     // Other plans are errors, and should be reported as such.
                     _ => return Err(()),
                 };
+                let mut required_arrangements = vec![HashSet::new(); inputs.len()];
+                // Delta joins should only be planned if a particular set of arrangements exists.
+                // Empirically, we've found that there are sometimes bugs causing them to be planned
+                // anyway. If this is the case, we need to create the arrangements, so we do so here,
+                // and complain with an error message.
+                if let JoinPlan::Delta(DeltaJoinPlan { path_plans }) = &plan {
+                    for DeltaPathPlan { stage_plans, .. } in path_plans {
+                        for DeltaStagePlan {
+                            lookup_relation,
+                            lookup_key,
+                            ..
+                        } in stage_plans
+                        {
+                            required_arrangements[*lookup_relation].insert(lookup_key.clone());
+                        }
+                    }
+                } else {
+                    // Linear joins handle rendering all the arrangements they need.
+                }
+                for (((arrangements, plan), required_arrangements), &arity) in input_keys
+                    .iter()
+                    .zip(plans.iter_mut())
+                    .zip(required_arrangements.iter())
+                    .zip(input_arities.iter())
+                {
+                    let missing: Vec<_> = required_arrangements
+                        .difference(arrangements)
+                        .cloned()
+                        .collect();
+                    if !missing.is_empty() {
+                        log::error!("Arrangements depended on by delta join alarmingly absent: {:?}
+This is not expected to cause incorrect results, but could indicate a performance issue in Materialize.", missing);
+                        let base_plan =
+                            std::mem::replace(plan, Plan::Constant { rows: Ok(vec![]) });
```

Comment:
I don't understand this. Don't we end up deleting the input and replacing it with the empty collection?

Reply:
Oh, wait. No. I understand how `replace` works now, sorry.